### ISSUE TITLE
stellarsolver combines incompatible GPL licenses (so was rejected from Debian)

### ISSUE THREAD

**Issue by rikmills** [rlancaste/stellarsolver#76]
https://lists.debian.org/debian-qt-kde/2021/02/msg00114.html

> in this package there are files under the GPL-2 (stellarsolver/astrometry/libkd/an-fls.h) mixed with files under GPL-3+.
> According to [1] this is not allowed.
> 
>   Thorsten
> 
> [1] https://www.gnu.org/licenses/gpl-faq.en.html#AllCompatibility

Also see: https://www.gnu.org/licenses/gpl-faq.en.html#v2v3Compatibility

**Comment by rlancaste** [comment]
Hi, I was not unaware of any issues with the licensing.  This project is based on code from astrometry.net and was produced with their feedback as well.  Astrometry.net includes the libkd code and their license is as follows:

"Parts of the code written by the Astrometry.net Team are licensed under a 3-clause BSD-style license. See the file LICENSE for the full license text. However, since this code uses libraries licensed under the GNU GPL, the whole work is distributed under the GPL version 3 or later."

https://github.com/dstndstn/astrometry.net

The libkd code was also released by them and is part of the project.  So is this a problem that they would need to correct?

https://github.com/dstndstn/astrometry.net/blob/master/libkd/LICENSE
https://github.com/dstndstn/astrometry.net/blob/master/libkd/README

**Comment by rikmills** [comment]
I guess it would, but could not be 100% simple to relicense to be compatible, as libkd/an-fls.h is licenced explicitly as:

> /**
>  This file includes code copied from three source files in the Linux kernel
>  v 2.6.23.1:
>  - include/asm-generic/bitops/fls.h
>  - include/asm-i386/bitops.h
>  - include/asm-x86_64/bitops.h
>  The second and third files contain this notice:
>  * Copyright 1992, Linus Torvalds.
>  And of course the kernel is distributed under the terms of the GPL v2.
>  */

However, I will raise an issue with astrometry shortly to see.

**Comment by pinotree** [comment]
This is also a problem because you are copying sources from another project, and this is exactly one of the downsides of it.
As I wrote in dstndstn/astrometry.net#212, IANAL however it seems to be the resulting binary (the stellarsolver library in this case) cannot be legally distributed.

It would be really nice to have some sort of cooperation with the astrometry.net project, so they can provide some public library that you can wrap in your StellarSolver, rather than embedding a copy of it (with all the downsides it implies).

**Comment by jriddell** [comment]
Yes I concur it is not legal to distribute libstellarsolver while it contains code which is both GPL 2 only and GPL 3.  an-fls.h needs rewritten or remove or asking Linus to licence as GPL 3+

**Comment by rikmills** [comment]
Debian is now in soft freeze, so stellarsolver will not get into the next stable release (release) as I read the policy. Ubuntu 21.04 is not yet in feature freeze, so there is still a (slim) chance there if we have a new release where it is clear the issues are fixed.

**Comment by rlancaste** [comment]
I am not sure if you saw the discussions on this page recently including my last post: https://github.com/dstndstn/astrometry.net/issues/212

Dustin made a very quick fix based on your posts and updated the code hopefully fixing the astrometry.net licensing issue.  I can fix it in StellarSolver this weekend.  But I think this issue affects much more than just StellarSolver if I understand correctly.  I can make changes to StellarSolver and KStars and I can contact several others people that I know package up astrometry.net.  I haven't gotten a response to my last post.

But for example, does this issue not cause a serious issue for this package that is already included in Debian?

https://packages.debian.org/sid/astrometry.net

It should also affect homebrew and Arch, as well as programs like Cloudmakers Astrometry Mac program.  So I can try to let some folks know, but I don't know everybody who needs to be told to update.

**Comment by rikmills** [comment]
I assume so (but IANAL either). I was planning to ping the Ubuntu release team, and perhaps file some bugs (debian/ubuntu), but as always spare time is a limiting factor.

**Comment by rlancaste** [comment]
I totally understand the spare time issue, as I said in the other thread, I have been really busy recently.  But this is a 3 day weekend now, and I should be able to at least fix the issue in StellarSolver based on what Dustin Lang fixed and hopefully rebuild KStars for Mac with the new updated library.  Then I can try to let some other folks know

**Comment by rlancaste** [comment]
Ok so I think I have updated the necessary files in StellarSolver to update the astrometry.net code and license with the update Dustin Lang made.  If this fixes the issue, I can update to a new version and make a new release.  But before I do that, please let me know if you think the issue has been fixed.

https://github.com/rlancaste/stellarsolver/commit/820befc37cd84d2a79135ee538c403fe34490de0

Thanks!

Rob

**Comment by rikmills** [comment]
@jriddell @pinotree Do you now think this looks ok?
